targets/p5.js

```diff
@@ -0,0 +1,188 @@
+
+  worldPos = vec4(aPosition, 1.);
+  sculptureCenter = vec3(0.);
+  cameraPosition = (uInverseModelViewMatrix * vec4(0., 0., 0., 1.0)).xyz;
+
+  gl_Position = uProjectionMatrix * mvPosition;
+}
+`
+
+  let generatedGLSL = sculptToGLSL(source);
+  let frag = `#version 300 es
+precision highp float;
+uniform mat4 uModelViewMatrix;
+uniform mat4 uProjectionMatrix;
+in vec4 worldPos;
+in vec3 sculptureCenter;
+in vec3 cameraPosition;
+out vec4 pc_fragColor;
+` +
+    usePBRHeader +
+    useHemisphereLight +
+    uniformsToGLSL(generatedGLSL.uniforms) +
+    "const float STEP_SIZE_CONSTANT = " +
+    generatedGLSL.stepSizeConstant +
+    ";\n" +
+    "const int MAX_ITERATIONS = " +
+    generatedGLSL.maxIterations +
+    ";\n" +
+    "#define MAX_REFLECTIONS " +
+    generatedGLSL.maxReflections +
+    "\n" + 
+    sculptureStarterCode.replace(/max_dist = [\d\.]+;/, 'max_dist = 1000.0;') +
+    generatedGLSL.geoGLSL +
+    "\n" +
+    generatedGLSL.colorGLSL +
+    "\n" +
+    fragFooter.replace(
+      '#ifdef MAX_REFLECTIONS',
+      `vec3 surface = rayOrigin + rayDirection*t;
+      #ifdef MAX_REFLECTIONS`
+    ).replace(
+      'pc_fragColor = vec4(outputColor, opacity);',
+      `pc_fragColor = vec4(outputColor, opacity);
+
+      // Output the correct depth value (in the 0-1 range) so that regular
+      // p5 3D objects interact in space correctly with Shader Park objects
+      vec4 screen = uProjectionMatrix * uModelViewMatrix * vec4(surface, 1.);
+      vec3 scaledScreen = screen.xyz / screen.w;
+      gl_FragDepth = scaledScreen.z * 0.5 + 0.5;`
+    );
+
+  return { vert, frag };
+}
+
+function createRenderer(target, {
+  scale = 1,
+  drawGeometry = (() => target.sphere(100)),
+} = {}) {
+  const output = {
+    shader: undefined,
+    draw: () => {
+      if (target.webglVersion !== target.WEBGL2) {
+        throw new Error(
+          'This p5 context uses WebGL1. Please use Shader Park on a WebGL2 context.'
+        );
+      }
+
+      const inverseModelViewMatrix =
+        new p5.Matrix().invert(_renderer.uMVMatrix)
+      const gl = window._renderer.GL;
+      gl.enable(gl.CULL_FACE);
+      gl.cullFace(gl.FRONT);
+
+      target.push();
+      target.noStroke();
+      target.shader(output.shader);
+      output.shader.setUniform(
+        'uInverseModelViewMatrix',
+        inverseModelViewMatrix.mat4
+      );
+      output.shader.setUniform('time', millis() / 1000);
+      output.shader.setUniform('opacity', 1);
+      output.shader.setUniform(
+        'mouse',
+        [
+          target.map(target.mouseX, 0, target.width, -1, 1),
+          target.map(target.mouseY, 0, target.height, 1, -1),
+          -0.5
+        ]
+      );
+      output.shader.setUniform('_scale', scale);
+      output.shader.setUniform('stepSize', 0.85);
+      output.shader.setUniform(
+        'resolution',
+        [width * target.pixelDensity(), height * target.pixelDensity()]
+      );
+      drawGeometry();
+      gl.disable(gl.CULL_FACE);
```

Comment:
Could this disable face culling when a p5 user wants it to stay on? (even if it did I wouldn't imagine it would affect many users.)

Reply:
good call, I'll grab the current parameter values at the start and restore them here